mlflow/store/dbmodels/models.py

```diff
@@ -44,7 +44,10 @@ def _create_entity(base, model):
                     # only get latest recorded metrics per key
                     metrics = {}
                     for o in obj:
-                        if o.key not in metrics or o.timestamp > metrics.get(o.key).timestamp:
+                        existing_metric = metrics.get(o.key)
+                        if existing_metric is None or (
+                                o.timestamp >= existing_metric.timestamp
```

Comment:
If we're looping over metrics that look like (pseudo-code):

```python
[
    Metric(name="foo", timestamp=10, value=5.0),
    Metric(name="foo", timestamp=12, value=2.0)
]
```

I think the first metric will be returned rather than the second, which I believe is not correct. This is because on the second pass through the loop, the objects will have the values:

```python
o.timestamp = 12
o.value = 2.0
existing_metric.timestamp = 10
existing_metric.timestamp = 5.0
```

And so the second part of the `if` statement will fail.

Would it be possible to perform some similar logic to that discussed in #999, for example by grouping first by metric name, then doing using `max`:

```python
groups = defaultdict(list)
for metric in obj:
    groups[metric.key].append(metric)
latest_metrics = [
    max(group, key=lambda m: (m.timestamp, m.value))
    for group in groups.values()
]
```

Reply:
Thanks for catching this. I've incorporated your suggestion and updated the test case to vary over multiple timestamps and values

Reply:
After discussion with @mparkhe, we'd prefer to fix the set of predicates by adding an additional `if o.timestamp > existing_metric.timestamp` check instead. While the code may not look as clean, it's more easily accessible and does not require us to allocate additional memory to store all metrics in a dictionary.

Reply:
Well as long as the logic is correct that's the important thing.

Ultimately if you're concerned about memory the 'right' solution to this (that avoids the need to load all metric values for the run into memory in Python in the first place) seems like it would be to encode this logic into the SQL query itself, though I'm not sufficiently experienced with the ORM to know best to achieve that.